puzzle-utils/words/dropquote.py

```diff
@@ -0,0 +1,193 @@
+from .search import search
+
+class DropQuoteSnapshot:
+    def __init__(self):
+        self.grid = []
+        self.letters = []
+        self.words = {}
+    
+    def print(self):
+        print(self.grid)
+        print(self.letters)
+        for index, words in self.words.items():
+            print(f'{index}: {words.letters} | {words.words}')
+        print()
+
+class DropQuoteWords:
+    def __init__(self, letters, words):
+        self.letters = letters
+        self.words = words
+
+def dropquote(grid, letters):
+    """
+    This solver will attempt to solve the drop quote puzzle and returns a set of snapshots on the iterations performed.
+    >>> grid = ['    -', '  -  ', '     ']
+    >>> letters = ['gor', 'afo', 'mn', 'ete', 'hs']
+    >>> snapshots = dropquote(grid, letters)
+    >>> print(snapshots[0].grid) # prints the initial state of the grid
+    ['    -', '  -  ', '     ']
+    >>> print(snapshots[0].letters) # prints the initial state of available letters
+    ['gor', 'afo', 'mn', 'ete', 'hs']
+    >>> for index, words in snapshots[0].words.items():
+    >>>     print(f'{index}: {words.letters} | {words.words}') # prints the index of the word, available letters, and found words
+    0: ['gor', 'afo', 'mn', 'ete'] | ['gant', 'rone', 'ront', 'oont', 'gone', 'rant', 'game', 'gane']
+    1: ['gor', 'afo'] | ['go', 'of']
+    2: ['ete', 'hs', 'gor', 'afo', 'mn', 'ete', 'hs'] | ['thrones']
+    >>> print(snapshots[-1].grid) # prints the current state of the grid
+    ['game-', 'of-th', 'rones']
+    """
+    dropquote_validate(grid, letters)
+
+    current_grid = []
+    for row in grid:
+        current_grid.append(row.lower())
+
+    current_letters = []
+    for column in letters:
+        current_letters.append(column.lower())
+
+    stuck = False
+    dropQuoteSnapshots = []
+    while not stuck:
+        stuck = True
+        dropQuoteSnapshot = dropquote_search(current_grid, current_letters)
+        dropQuoteSnapshots.append(dropQuoteSnapshot)
+
+        for index, words in dropQuoteSnapshot.words.items():
+            if len(words.words) == 1:
+                stuck = False
+                dropquote_apply_word(current_grid, current_letters, index, words.words[0])
+    
+    return dropQuoteSnapshots
+
+def dropquote_validate(grid, letters):
+    row_length = len(grid[0])
+    column_lengths = [0] * row_length
+    for row in range(len(grid)):
```

Comment:
You get a less verbose error message, but a one-liner for this can be something like:
```
if not all(map(lambda row: row_length == len(row), grid)):
    <error>
```